### ISSUE TITLE
answer normalization

### ISSUE THREAD

**Issue by elbaro** [Cyanogenoid/pytorch-vqa#5]
The answer normalization code is different from the official VQA code.

1. Can we assume this?
"normalizations is not needed, assuming that the human answers are already normalized."

2. The official code removes the articles (a, the) but this code doesn't.
There are actually "the" in answers.

**Comment by Cyanogenoid** [comment]
You are correct, my evaluation code is slightly different. After looking into this, it seems to me that their [evaluation specification](http://visualqa.org/evaluation.html) is badly designed.

As far as I can tell, in the official code, machine-given answers are normalized in [several ways](https://github.com/GT-Vision-Lab/VQA/blob/master/PythonEvaluationTools/vqaEvaluation/vqaEval.py#L87-L91) -- such as the removal of articles as you mention -- but the ground-truth answers it checks against only have their punctuation normalized. I made the assumption that you mention based on the fact that their evaluation would clearly be wrong if the ground truth is not normalized: if an unnormalized ground truth answer would be changed due to normalization if it were given as a machine answer, it is impossible for any machine-given answer to be considered equal to this ground truth answer. For example, if all ground truth answers are "the cat" and the machine predicts "the cat", then normalization will only change the machine answer to "cat", which is unequal to "the cat" and so receive a score of 0 in the official code.

As you correctly point out, the ground truth answers are indeed not normalized. This should mean that my code evaluates slightly more things to be correct (the ones where the normalization doesn't make a correct answer wrong) and that it might be worth filing an issue in their repository. A quick search through the data suggests that about 1750 questions in the training set are affected by this (most of them are answers that contain the word "one"). I would rather not introduce this "bug" into my code to make it equal to theirs.

**Comment by elbaro** [comment]
I agree the official method is broken. In my guess the original paper still used the old evaluation method, otherwise it cannot be compared to other papers.

You might want to clarify in README the list of differences to the paper (or any nontrivial assumption).

**Comment by Cyanogenoid** [comment]
I'll do that, thanks for flagging up these issues!
